src/strongvelope/strongvelope.cpp

```diff
@@ -738,18 +805,50 @@ void ProtocolHandler::rsaDecrypt(const StaticBuffer& data, Buffer& output)
 }
 
 promise::Promise<std::pair<MsgCommand*, KeyCommand*>>
-ProtocolHandler::msgEncrypt(Message* msg, MsgCommand* msgCmd)
+ProtocolHandler::msgEncrypt(Message* msg, const SetOfIds &recipients, MsgCommand* msgCmd)
 {
     assert(msg->keyid == msgCmd->keyId());
-    if ((msg->keyid == CHATD_KEYID_INVALID)
-     || (msg->keyid == CHATD_KEYID_UNCONFIRMED)) //we have to use the current send key
+    if ((msg->keyid == CHATD_KEYID_INVALID)         // keyid has not been assigned yet (NEWMSG)
+     || (msg->keyid == CHATD_KEYID_UNCONFIRMED))    // or it was, but not yet confimed (MSGUPDX, replayed NEWMSG)
     {
-        if (!mCurrentKey || mParticipantsChanged) // create a new key and prepare the KeyCommand
+        assert(msgCmd->opcode() != OP_MSGUPD);  // edits always have a confirmed keyid
+
+        if (mCurrentKey && mCurrentKeyParticipants == *mParticipants)
+        {
+            // use current key, whether it's unconfirmed (in-flight) or already confirmed
+            assert(mCurrentKey);
+            assert(mCurrentKeyId != CHATD_KEYID_INVALID);
+            msg->keyid = mCurrentKeyId;
+            msgCmd->setKeyId(mCurrentKeyId);
+            msgEncryptWithKey(*msg, *msgCmd, *mCurrentKey);
+            return std::make_pair(msgCmd, (KeyCommand*)nullptr);
+        }
+        else    // msg for different set of recipients or no current key available
         {
+            // edits of unconfirmed new messages should reuse the same key than original message
+            if (msgCmd->opcode() == OP_MSGUPDX)
+            {
+                std::shared_ptr<SendKey> unconfirmedKey;
+                for (auto& it: mUnconfirmedKeys)
+                {
+                    if (it.first == recipients)
+                    {
+                        unconfirmedKey = it.second;
+
+                        msg->keyid = CHATD_KEYID_UNCONFIRMED;
+                        msgCmd->setKeyId(CHATD_KEYID_UNCONFIRMED);
+                        msgEncryptWithKey(*msg, *msgCmd, *unconfirmedKey);
+
+                        return std::make_pair(msgCmd, (KeyCommand*)nullptr);
+                    }
+                }
+                return promise::Error("Unconfirmed key for edit of msg "+msg->id().toString()+" not found.", EINVAL, SVCRYPTO_ENOKEY);
```

Comment:
Review that a failed promise here is handled at a higher level.

Reply:
The `.fail()` handler is at the chat-client. As result of a failed promise, it logs the error and discards the new message.